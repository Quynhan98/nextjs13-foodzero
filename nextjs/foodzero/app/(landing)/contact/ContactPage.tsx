'use client'

import { useState, useCallback, FormEvent, ChangeEvent } from 'react'
import { Box, useToast } from '@chakra-ui/react'

// Constants
import {
  COUNTRY_CODE,
  INVALID_EMAIL_FORMAT,
  NUMBER_OF_PERSON,
  PHONE_ERROR_LENGTH,
  REGEX_EMAIL,
  REGEX_REMOVE_BRACKETS,
  RESERVATION_TIME,
  SERVER_ERROR,
  SNACKBAR_BOOKING_SUCCESS,
  TIME_DUPLICATED_ERROR,
} from '@constants/index'

// Components
import ReservationForm from '@components/ReservationForm'
import LoadingIndicator from '@components/LoadingIndicator'
import Address from '@components/Sections/Address'
import HeroContactPage from '@components/Sections/HeroContactPage'

// Hooks
import { useBookingContext } from '@hooks/useBookingContext'
import { useLoadingContext } from '@hooks/useLoadingContext'

// Context
import { IBookingContext } from '@contexts/BookingProvider'

// Utils
import { findItemByValue, formatDate, formatPhoneNumber } from '@utils/index'
import { checkValidate } from '@utils/validation'

const reservationInit = {
  date: '' as unknown as Date,
  time: RESERVATION_TIME[0],
  person: NUMBER_OF_PERSON[0],
  phone: '',
  firstName: '',
  lastName: '',
  email: '',
}

const initErrorMsgs = {
  email: '',
  phone: '',
  time: '',
}

const ContactPage = () => {
  const toast = useToast()
  const { booking, addBooking } = useBookingContext() as IBookingContext
  const { setLoading, loading } = useLoadingContext()
  const [reservation, setReservation] = useState(reservationInit)
  const [errorMessage, setErrorMessage] = useState(initErrorMsgs)

  // Check disable button
  const isDisable = !!(
    errorMessage.email ||
    errorMessage.phone ||
    errorMessage.time ||
    !reservation.date
  )

  const handleChangeDate = useCallback(
    (date: Date) => {
      const newDate = date && formatDate(date)
      const dateDuplicated = booking.filter(
        (item) => formatDate(new Date(item.date)) === newDate,
      )

      if (dateDuplicated.length > 0) {
        const timeDuplicated = findItemByValue({
          data: dateDuplicated,
          value: reservation.time,
          key: 'time',
        })

        setErrorMessage((prev) => ({
          ...prev,
          time: timeDuplicated ? TIME_DUPLICATED_ERROR : '',
        }))
      } else {
        setErrorMessage((prev) => ({
          ...prev,
          time: '',
        }))
      }

      setReservation((prev) => ({
        ...prev,
        date,
      }))
    },
    [booking, reservation.time],
  )

  const handleChangeField = useCallback(
    (e: ChangeEvent<HTMLInputElement | HTMLSelectElement>) => {
      const nameInput = e.target.name
      const valueInput = e.target.value
      const dateDuplicated =
        reservation.date &&
        booking.filter(
          (item) =>
            formatDate(new Date(item.date)) === formatDate(reservation.date),
        )

      if (nameInput === 'email') {
        const validateEmail = checkValidate({
          value: valueInput,
          regex: REGEX_EMAIL,
          errorMess: INVALID_EMAIL_FORMAT,
        })

        setErrorMessage((prev) => ({
          ...prev,
          email: validateEmail || '',
        }))
      }

      if (nameInput === 'phone') {
        const phone = formatPhoneNumber(valueInput).replace(
          REGEX_REMOVE_BRACKETS,
          '',
        )

        setErrorMessage((prev) => ({
          ...prev,
          phone: phone.length !== 10 ? PHONE_ERROR_LENGTH : '',
        }))
      }

      if (dateDuplicated.length > 0 && nameInput === 'time') {
        const timeDuplicated = findItemByValue({
          data: dateDuplicated,
          value: valueInput,
          key: 'time',
        })

        setErrorMessage((prev) => ({
          ...prev,
          time: timeDuplicated ? TIME_DUPLICATED_ERROR : '',
        }))
      }

      setReservation((prev) => ({
        ...prev,
        [nameInput]: valueInput,
      }))
    },
    [booking, reservation.date],
  )

  const handleSubmit = useCallback(
    async (e: FormEvent<HTMLFormElement>) => {
      e?.preventDefault()

      const payload = {
        ...reservation,
        phone: reservation.phone
          ? COUNTRY_CODE +
            formatPhoneNumber(reservation.phone).replace(
              REGEX_REMOVE_BRACKETS,
              '',
            )
          : '',
      }

      try {
        setLoading(true)
        await addBooking([...booking, payload])

        toast({
          title: 'Success',
          description: SNACKBAR_BOOKING_SUCCESS,
          status: 'success',
          isClosable: true,
          position: 'bottom-left',
        })

        setReservation(reservationInit)
      } catch (error) {
        toast({
          title: 'Error',
          description: SERVER_ERROR,
          status: 'error',
          isClosable: true,
          position: 'bottom-left',
        })
      } finally {
        setLoading(false)
      }
    },
    [booking, reservation],
  )

  return (
    <>
      {/* Hero Section */}
      <HeroContactPage />

      {/* Address section */}
      <Address />

      {/* Reservation Section */}
      <Box
        id="reservationSection"
        as="section"
        padding={{
          base: '70px 12px',
          md: '180px 50px',
          '2xl': '237px 138px 254px 138px',
        }}
        backgroundColor="alabaster"
        position="relative"
        _before={{
          content: '""',
          position: 'absolute',
          width: { base: '153px', md: '358px', '2xl': '511px' },
          height: { base: '104px', md: '243px', '2xl': '347px' },
          top: { base: '15px', md: '50px' },
          left: { base: '18px', md: '50px', '2xl': '159px' },
          backgroundImage: '/images/mint.webp',
          backgroundRepeat: 'no-repeat',
          backgroundSize: 'cover',
          zIndex: 0,
        }}
      >
        <ReservationForm
          {...reservation}
          timeError={errorMessage.time}
          emailError={errorMessage.email}
          phoneError={errorMessage.phone}
          isShowFullField
          phone={formatPhoneNumber(reservation.phone)}
          onSubmitForm={handleSubmit}
          handleChangeDate={handleChangeDate}
          onChangeField={handleChangeField}
          isDisableButton={isDisable}
        />
      </Box>
      {loading && <LoadingIndicator size="lg" />}
    </>
  )
}

export default ContactPage
