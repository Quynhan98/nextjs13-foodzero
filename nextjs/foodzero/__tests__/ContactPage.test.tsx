// Utils
import {
  cleanup,
  screen,
  render,
  RenderResult,
  fireEvent,
  waitFor,
} from '@utils/testUtils'
import { formatPhoneNumber } from '@utils/index'

// Constants
import {
  INVALID_EMAIL_FORMAT,
  PHONE_ERROR_LENGTH,
  SERVER_ERROR,
  SNACKBAR_BOOKING_SUCCESS,
} from '@constants/index'

// App
import ContactPage from 'app/(landing)/contact/ContactPage'

describe('Contact page', () => {
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve(),
    }),
  ) as jest.Mock

  const bookingMock = {
    firstName: 'John',
    lastName: 'Wick',
    email: 'John@gmail.com',
    phone: '1231231231',
    date: '12/21/2022',
    time: '12:00 am',
    person: '2 Person',
  }

  let wrapper: RenderResult

  beforeEach(() => {
    wrapper = render(<ContactPage />)
  })

  afterEach(() => {
    cleanup()
  })

  it('Renders Contact page correctly', () => {
    const { container } = wrapper

    expect(container).toMatchSnapshot()
  })

  it('Should call onChange first name', () => {
    const firstName = screen.getByPlaceholderText('First Name')

    fireEvent.change(firstName, {
      target: { value: bookingMock.firstName },
    })

    expect(firstName).toHaveValue(bookingMock.firstName)
  })

  it('Should call onChange last name', () => {
    const lastName = screen.getByPlaceholderText('Last Name')

    fireEvent.change(lastName, {
      target: { value: bookingMock.lastName },
    })

    expect(lastName).toHaveValue(bookingMock.lastName)
  })

  it('Should call onChange email', () => {
    const email = screen.getByPlaceholderText('Email')

    fireEvent.change(email, {
      target: { value: bookingMock.email },
    })

    expect(email).toHaveValue(bookingMock.email)
  })

  it('Should call onChange phone', () => {
    const phone = screen.getByPlaceholderText('Phone')

    fireEvent.change(phone, {
      target: { value: bookingMock.phone },
    })

    expect(phone).toHaveValue(formatPhoneNumber(bookingMock.phone))
  })

  it('Should call onChange date picker', () => {
    const date = screen.getByPlaceholderText('MM/DD/YYYY')

    fireEvent.change(date, {
      target: { value: bookingMock.date },
    })

    expect(date).toHaveValue(bookingMock.date)
  })

  it('Should onchange time ', () => {
    const time = screen.getByTestId('selectTime')

    fireEvent.change(time, { target: { value: bookingMock.time } })

    expect(time).toHaveValue(bookingMock.time)
  })

  it('Should onchange person ', () => {
    const person = screen.getByTestId('selectPerson')

    fireEvent.change(person, { target: { value: bookingMock.person } })

    expect(person).toHaveValue(bookingMock.person)
  })

  it('Should call submit form booking', () => {
    const button = screen.getByRole('button', { name: /Book Now/i })

    fireEvent.submit(button)

    waitFor(() => {
      expect(screen.getByText(SNACKBAR_BOOKING_SUCCESS)).toBeTruthy()
    })
  })

  it('Should call submit form booking with server error', () => {
    jest.fn().mockImplementationOnce(() => Promise.reject(SERVER_ERROR))

    const button = screen.getByRole('button', { name: /Book Now/i })

    fireEvent.submit(button)

    waitFor(() => {
      expect(screen.getByText(SERVER_ERROR)).toBeTruthy()
    })
  })

  it('Should call submit form booking with format email and phone error', () => {
    const email = screen.getByPlaceholderText('Email')
    const phone = screen.getByPlaceholderText('Phone')
    const button = screen.getByRole('button', { name: /Book Now/i })

    fireEvent.change(email, { target: { value: 'notEmail' } })
    fireEvent.change(phone, { target: { value: '00' } })
    fireEvent.submit(button)

    waitFor(() => {
      expect(screen.getByText(PHONE_ERROR_LENGTH)).toBeTruthy()
      expect(screen.getByText(INVALID_EMAIL_FORMAT)).toBeTruthy()
    })
  })
})
