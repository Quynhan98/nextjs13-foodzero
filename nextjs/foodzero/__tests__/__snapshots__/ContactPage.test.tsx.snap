// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact page Renders Contact page correctly 1`] = `
<div>
  <section
    class="css-hviug4"
  >
    <img
      alt="background contact page"
      data-nimg="fill"
      decoding="async"
      sizes="100vw"
      src="/_next/image?url=%2Fimages%2FcontactBackground.webp&w=3840&q=75"
      srcset="/_next/image?url=%2Fimages%2FcontactBackground.webp&w=640&q=75 640w, /_next/image?url=%2Fimages%2FcontactBackground.webp&w=750&q=75 750w, /_next/image?url=%2Fimages%2FcontactBackground.webp&w=828&q=75 828w, /_next/image?url=%2Fimages%2FcontactBackground.webp&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2FcontactBackground.webp&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2FcontactBackground.webp&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2FcontactBackground.webp&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2FcontactBackground.webp&w=3840&q=75 3840w"
      style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; z-index: -1; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
    />
    <h2
      class="chakra-heading css-idxew5"
    >
      Get in Touch
    </h2>
    <p
      class="chakra-text css-4odzap"
    >
      The freshest ingredients for you every day
    </p>
    <div
      class="css-kns484"
    >
      <div
        class="css-lex2vo"
      >
        <div
          class="css-1rdnf7o"
        >
          <p
            class="chakra-text css-1j8yc53"
          >
            Open Time
          </p>
          <p
            class="chakra-text css-1nxiabe"
          >
            Sunday - Friday
          </p>
        </div>
        <hr
          class="css-1iijbtl"
        />
        <ul
          class="css-y465uq"
          role="list"
        >
          <li
            class="css-0"
          >
            <p
              class="chakra-text css-1nxiabe"
            >
              Brunch
            </p>
            <p
              class="chakra-text css-1nxiabe"
            >
              11:00–12:00
            </p>
          </li>
          <li
            class="css-0"
          >
            <p
              class="chakra-text css-1nxiabe"
            >
              Lunch
            </p>
            <p
              class="chakra-text css-1nxiabe"
            >
              13:00–17:00
            </p>
          </li>
          <li
            class="css-0"
          >
            <p
              class="chakra-text css-1nxiabe"
            >
              Dinner
            </p>
            <p
              class="chakra-text css-1nxiabe"
            >
              18:00–20:00
            </p>
          </li>
        </ul>
      </div>
    </div>
  </section>
  <section
    class="css-25f137"
  >
    <div
      class="css-uktyfy"
    >
      <div
        class="css-314wmh"
      >
        <img
          alt="excellent cook picture"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          sizes="(max-width: 768px) 277px, 208px (min-width: 768px) 544px, 415px (min-width: 1200px) 792px, 593px"
          src="/_next/image?url=%2Fimages%2FdinnerTable.webp&w=3840&q=75"
          srcset="/_next/image?url=%2Fimages%2FdinnerTable.webp&w=16&q=75 16w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=32&q=75 32w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=48&q=75 48w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=64&q=75 64w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=96&q=75 96w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=128&q=75 128w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=256&q=75 256w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=384&q=75 384w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=640&q=75 640w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=750&q=75 750w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=828&q=75 828w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2FdinnerTable.webp&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
        />
      </div>
      <p
        class="chakra-text css-19q5b95"
      >
        We can be contacted 
        <br />
         via email
         
        <a
          href="mailto:info@foodzero.com"
          style="text-decoration: none;"
        >
          <b
            class="chakra-text css-ycbatn"
          >
            info@foodzero.com
          </b>
        </a>
        <br />
        or telephone on
         
        <a
          href="tel:+86 852 346 000"
          style="text-decoration: none;"
        >
          <b
            class="chakra-text css-ycbatn"
          >
            +86 852 346 000
          </b>
        </a>
      </p>
    </div>
    <div
      class="css-6mdup3"
    >
      <p
        class="chakra-text css-shftav"
      >
        We are located in 
        1959 Sepulveda Blvd. Culver City, CA, 90230
      </p>
      <div
        class="css-1j5gmxr"
      >
        <img
          alt="excellent cook picture"
          data-nimg="fill"
          decoding="async"
          loading="lazy"
          sizes="(max-width: 768px) 277px, 347px (min-width: 768px) 554px, 630px (min-width: 1200px) 792px, 990px"
          src="/_next/image?url=%2Fimages%2Faddress.webp&w=3840&q=75"
          srcset="/_next/image?url=%2Fimages%2Faddress.webp&w=16&q=75 16w, /_next/image?url=%2Fimages%2Faddress.webp&w=32&q=75 32w, /_next/image?url=%2Fimages%2Faddress.webp&w=48&q=75 48w, /_next/image?url=%2Fimages%2Faddress.webp&w=64&q=75 64w, /_next/image?url=%2Fimages%2Faddress.webp&w=96&q=75 96w, /_next/image?url=%2Fimages%2Faddress.webp&w=128&q=75 128w, /_next/image?url=%2Fimages%2Faddress.webp&w=256&q=75 256w, /_next/image?url=%2Fimages%2Faddress.webp&w=384&q=75 384w, /_next/image?url=%2Fimages%2Faddress.webp&w=640&q=75 640w, /_next/image?url=%2Fimages%2Faddress.webp&w=750&q=75 750w, /_next/image?url=%2Fimages%2Faddress.webp&w=828&q=75 828w, /_next/image?url=%2Fimages%2Faddress.webp&w=1080&q=75 1080w, /_next/image?url=%2Fimages%2Faddress.webp&w=1200&q=75 1200w, /_next/image?url=%2Fimages%2Faddress.webp&w=1920&q=75 1920w, /_next/image?url=%2Fimages%2Faddress.webp&w=2048&q=75 2048w, /_next/image?url=%2Fimages%2Faddress.webp&w=3840&q=75 3840w"
          style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent; background-size: cover; background-position: 50% 50%; background-repeat: no-repeat;"
        />
      </div>
    </div>
  </section>
  <section
    class="css-2ce55m"
    id="reservationSection"
  >
    <form
      novalidate=""
    >
      <div
        class="chakra-form-control css-qjxwnb"
        role="group"
      >
        <div
          class="css-yjtq8f"
        >
          <h3
            class="chakra-heading css-17j5m7v"
          >
            Make a Reservation
          </h3>
          <p
            class="chakra-text css-1j19xza"
          >
            Get in touch with restaurant
          </p>
        </div>
        <div
          class="css-19580xy"
        >
          <input
            class="chakra-input css-sw9qjs"
            id="field-:r0:"
            name="firstName"
            placeholder="First Name"
            type="text"
            value=""
          />
          <input
            class="chakra-input css-sw9qjs"
            id="field-:r0:"
            name="lastName"
            placeholder="Last Name"
            type="text"
            value=""
          />
        </div>
        <div
          class="css-1qnzkxj"
        >
          <input
            class="chakra-input css-1njo91c"
            id="field-:r0:"
            name="email"
            placeholder="Email"
            type="email"
            value=""
          />
        </div>
        <div
          class="css-1qnzkxj"
        >
          <input
            class="chakra-input css-1njo91c"
            id="field-:r0:"
            name="phone"
            placeholder="Phone"
            type="tel"
            value=""
          />
        </div>
        <div
          class="css-r2rtav"
        >
          <div
            class="css-1payax8"
          >
            <div
              class="css-0"
            >
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    class="chakra-input css-1cwrfhe"
                    id="field-:r0:"
                    placeholder="MM/DD/YYYY"
                    value=""
                  />
                </div>
              </div>
            </div>
            <div
              class="css-0"
            >
              <div
                class="chakra-select__wrapper css-sp79o5"
              >
                <select
                  class="chakra-select css-lknr9b"
                  data-testid="selectTime"
                  id="field-:r0:"
                  name="time"
                  style="padding-left: 40px;"
                >
                  <option
                    value=""
                  >
                    11:00 am
                  </option>
                  <option
                    value="11:00 am"
                  >
                    11:00 am
                  </option>
                  <option
                    value="12:00 am"
                  >
                    12:00 am
                  </option>
                  <option
                    value="13:00 pm"
                  >
                    13:00 pm
                  </option>
                  <option
                    value="14:00 pm"
                  >
                    14:00 pm
                  </option>
                  <option
                    value="15:00 pm"
                  >
                    15:00 pm
                  </option>
                  <option
                    value="16:00 pm"
                  >
                    16:00 pm
                  </option>
                  <option
                    value="17:00 pm"
                  >
                    17:00 pm
                  </option>
                  <option
                    value="18:00 pm"
                  >
                    18:00 pm
                  </option>
                  <option
                    value="19:00 pm"
                  >
                    19:00 pm
                  </option>
                </select>
                <div
                  class="chakra-select__icon-wrapper css-1rn5cz7"
                >
                  <svg
                    aria-hidden="true"
                    class="chakra-icon chakra-select__icon css-m73ylg"
                    focusable="false"
                    role="presentation"
                    style="width: 1em; height: 1em; color: currentColor;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
          <div
            class="chakra-select__wrapper css-42b2qy"
          >
            <select
              class="chakra-select css-lknr9b"
              data-testid="selectPerson"
              id="field-:r0:"
              name="person"
              style="padding-left: 40px;"
            >
              <option
                value=""
              >
                1 Person
              </option>
              <option
                value="1 Person"
              >
                1 Person
              </option>
              <option
                value="2 Person"
              >
                2 Person
              </option>
              <option
                value="3 Person"
              >
                3 Person
              </option>
              <option
                value="4 Person"
              >
                4 Person
              </option>
              <option
                value="5 Person"
              >
                5 Person
              </option>
              <option
                value="6 Person"
              >
                6 Person
              </option>
              <option
                value="7 Person"
              >
                7 Person
              </option>
              <option
                value="8 Person"
              >
                8 Person
              </option>
              <option
                value="9 Person"
              >
                9 Person
              </option>
              <option
                value="10 Person"
              >
                10 Person
              </option>
            </select>
            <div
              class="chakra-select__icon-wrapper css-1rn5cz7"
            >
              <svg
                aria-hidden="true"
                class="chakra-icon chakra-select__icon css-m73ylg"
                focusable="false"
                role="presentation"
                style="width: 1em; height: 1em; color: currentColor;"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
        </div>
        <div
          class="css-ffarb6"
        >
          <button
            class="chakra-button css-mqrzow"
            disabled=""
            type="submit"
          >
            Book Now
          </button>
        </div>
      </div>
    </form>
  </section>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
